Tests fail with Python 3.12

Attempting a build against Python 3.12 b3 fails some tests:

```
======================================================================
ERROR: test_malformed_wkt_no_ending_paren (geomet.tests.wkt_test.GeometryCollectionLoadsTestCase.test_malformed_wkt_no_ending_paren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/geomet-1.0.0/geomet/tests/wkt_test.py", line 1219, in test_malformed_wkt_no_ending_paren
    wkt.loads(mp)
  File "/builddir/build/BUILD/geomet-1.0.0/geomet/wkt.py", line 147, in loads
    result = importer(tokens, string)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/builddir/build/BUILD/geomet-1.0.0/geomet/wkt.py", line 638, in _load_geometrycollection
    geom = load_func(tokens, string)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not callable
======================================================================
FAIL: test_raises_unmatched_paren (geomet.tests.wkt_test.LineStringLoadsTestCase.test_raises_unmatched_paren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/geomet-1.0.0/geomet/tests/wkt_test.py", line 324, in test_raises_unmatched_paren
    self.assertEqual('Invalid WKT: `LINESTRING (0.0 1.0`',
AssertionError: 'Invalid WKT: `LINESTRING (0.0 1.0`' != "could not convert string to float: ''"
- Invalid WKT: `LINESTRING (0.0 1.0`
+ could not convert string to float: ''
======================================================================
FAIL: test_malformed_wkt_misbalanced_parens (geomet.tests.wkt_test.MultiPointLoadsTestCase.test_malformed_wkt_misbalanced_parens)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/geomet-1.0.0/geomet/tests/wkt_test.py", line 521, in test_malformed_wkt_misbalanced_parens
    self.assertEqual(expected, str(ar.exception))
AssertionError: 'Invalid WKT: `MULTIPOINT ((0 0), (0 1)`' != "could not convert string to float: ''"
- Invalid WKT: `MULTIPOINT ((0 0), (0 1)`
+ could not convert string to float: ''
======================================================================
FAIL: test_raises_unmatched_paren (geomet.tests.wkt_test.PointLoadsTestCase.test_raises_unmatched_paren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/geomet-1.0.0/geomet/tests/wkt_test.py", line 245, in test_raises_unmatched_paren
    self.assertEqual('Invalid WKT: `POINT (0.0 1.0`',
AssertionError: 'Invalid WKT: `POINT (0.0 1.0`' != "could not convert string to float: ''"
- Invalid WKT: `POINT (0.0 1.0`
+ could not convert string to float: ''
======================================================================
FAIL: test_raises_unmatched_paren (geomet.tests.wkt_test.PolygonLoadsTestCase.test_raises_unmatched_paren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/geomet-1.0.0/geomet/tests/wkt_test.py", line 431, in test_raises_unmatched_paren
    self.assertEqual(
AssertionError: 'Invalid WKT: `POLYGON ((0.0 0.0, 1.0 4.0, 4.0 1.0, 0.0 0.0)`' != "could not convert string to float: ''"
- Invalid WKT: `POLYGON ((0.0 0.0, 1.0 4.0, 4.0 1.0, 0.0 0.0)`
+ could not convert string to float: ''
Could not read Geopackage geometry because of errors: Envelope indicator must be between 0-4
Could not read Geopackage geometry because of errors: Geopackage version must be 0
```

It appears that the convenient tokenizer has changed behaviour a bit.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -28,6 +28,11 @@ py311_container: &py311_container
         - image: python:3.11-slim
     resource_class: small
     working_directory: ~/project
+py312_container: &py312_container
+    docker:
+        - image: python:3.12-slim
+    resource_class: small
+    working_directory: ~/project
 
 filter_build_test_package: &filter_build_test_package
     filters:
@@ -111,6 +116,11 @@ jobs:
         steps:
             - build_test_package:
                 py_ver: "311"
+    build_test_package_py312:
+        <<: *py312_container
+        steps:
+            - build_test_package:
+                py_ver: "312"
     publish_to_test_pypi:
         <<: *py37_container
         steps:
@@ -153,6 +163,9 @@ workflows:
             - build_test_package_py311:
                 <<: *filter_build_test_package
                 requires: [code_checkout]
+            - build_test_package_py312:
+                <<: *filter_build_test_package
+                requires: [code_checkout]
     release:
         jobs:
             - code_checkout:
@@ -172,6 +185,9 @@ workflows:
             - build_test_package_py311:
                 <<: *filter_release
                 requires: [code_checkout]
+            - build_test_package_py312:
+                <<: *filter_release
+                requires: [code_checkout]
             - publish_to_test_pypi:
                 <<: *filter_release
                 context: "Test PyPI"
@@ -181,6 +197,7 @@ workflows:
                     - build_test_package_py39
                     - build_test_package_py310
                     - build_test_package_py311
+                    - build_test_package_py312
             - approve_publish:
                 <<: *filter_release
                 type: approval

--- a/geomet/wkt.py
+++ b/geomet/wkt.py
@@ -106,10 +106,7 @@ def loads(string):
     """
     Construct a GeoJSON `dict` from WKT (`string`).
     """
-    sio = io.StringIO(string)
-    # NOTE: This is not the intended purpose of `tokenize`, but it works.
-    tokens = (x[1] for x in tokenize.generate_tokens(sio.readline))
-    tokens = _tokenize_wkt(tokens)
+    tokens = _tokenize_wkt(string)
     geom_type_or_srid = next(tokens)
     srid = None
     geom_type = geom_type_or_srid
@@ -144,17 +141,25 @@ def loads(string):
     return result
 
 
-def _tokenize_wkt(tokens):
+def _tokenize_wkt(string):
     """
     Since the tokenizer treats "-" and numeric strings as separate values,
     combine them and yield them as a single token. This utility encapsulates
     parsing of negative numeric values from WKT can be used generically in all
     parsers.
     """
+    sio = io.StringIO(string)
+    # NOTE: This is not the intended purpose of `tokenize`, but it works.
+    tokens = (x[1] for x in tokenize.generate_tokens(sio.readline))
     negative = False
     for t in tokens:
         if t == '-':
             negative = True
+            continue
+        elif t == '':
+            # Ignore empty string tokens.
+            # This can happen in python3.12, seemingly due to
+            # https://peps.python.org/pep-0701/#changes-to-the-tokenize-module
             continue
         else:
             if negative:

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -37,6 +37,7 @@ classifiers = [
     "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
     "Topic :: Scientific/Engineering :: GIS",
 ]
 dependencies = [

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,6 @@
 [tox]
 minversion = 3.26.0
-envlist = py37,py38,py39,py310,py311,style
+envlist = py37,py38,py39,py310,py311,py312,style
 
 [testenv]
 install_command = pip install -U {opts} {packages}

